src/CausationMetadataEnricher.php

```diff
@@ -14,30 +14,35 @@
 
 use ArrayIterator;
 use Prooph\Common\Event\ActionEvent;
-use Prooph\Common\Event\ActionEventEmitter;
-use Prooph\Common\Event\ActionEventListenerAggregate;
-use Prooph\Common\Event\DetachAggregateHandlers;
 use Prooph\Common\Messaging\Message;
 use Prooph\EventStore\ActionEventEmitterEventStore;
 use Prooph\EventStore\Metadata\MetadataEnricher;
-use Prooph\EventStore\Plugin\Plugin;
+use Prooph\EventStore\Plugin\Plugin as EventStorePlugin;
 use Prooph\EventStore\Stream;
 use Prooph\ServiceBus\CommandBus;
+use Prooph\ServiceBus\MessageBus;
+use Prooph\ServiceBus\Plugin\Plugin as MessageBusPlugin;
 
-final class CausationMetadataEnricher implements ActionEventListenerAggregate, MetadataEnricher, Plugin
+final class CausationMetadataEnricher implements MetadataEnricher, EventStorePlugin, MessageBusPlugin
 {
-    use DetachAggregateHandlers;
-
     /**
      * @var Message
      */
     private $currentCommand;
 
-    public function setUp(ActionEventEmitterEventStore $eventStore): void
-    {
-        $eventEmitter = $eventStore->getActionEventEmitter();
+    /**
+     * @var array
+     */
+    private $eventStoreListeners = [];
 
-        $eventEmitter->attachListener(
+    /**
+     * @var array
+     */
+    private $messageBusListeners = [];
+
+    public function attachToEventStore(ActionEventEmitterEventStore $eventStore): void
+    {
+        $this->eventStoreListeners[] = $eventStore->attach(
             ActionEventEmitterEventStore::EVENT_APPEND_TO,
             function (ActionEvent $event): void {
                 if (null === $this->currentCommand || ! $this->currentCommand instanceof Message) {
```

Comment:
`null` is also NOT an instance of `Message`, so this is a doubled expression I guess. Not sure about performance here.